src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractTypeImportCompletionProvider.cs

```diff
@@ -42,32 +52,104 @@ protected override async Task AddCompletionItemsAsync(CompletionContext completi
+                dictionary.Add(namespaceOfTarget, typeNameOfTarget);
+            }
+
+            return dictionary;
+        }
+
+        private IEnumerable<ITypeSymbol> GetAliasSymbolTargets(
+            SyntaxContext syntaxContext,
+            CancellationToken cancellationToken)
+        {
+            // In case the caret is at the beginning of the file, take the root node.
+            var aliasDeclarations = GetAliasDeclarationNodes(
+                syntaxContext.LeftToken.Parent ?? syntaxContext.SyntaxTree.GetRoot(cancellationToken));
+            foreach (var aliasNode in aliasDeclarations)
+            {
+                var symbol = syntaxContext.SemanticModel.GetDeclaredSymbol(aliasNode, cancellationToken);
+                if (symbol is IAliasSymbol {Target: ITypeSymbol target})
+                {
+                    yield return target;
+                }
+            }
+        }
+
+        private static void AddItems(
+            ImmutableArray<CompletionItem> items,
+            CompletionContext completionContext,
+            HashSet<string> namespacesInScope,
+            MultiDictionary<string, string> aliasTargetNamespaceToTypeNameMap,
+            TelemetryCounter counter)
         {
             counter.ReferenceCount++;
             foreach (var item in items)
             {
-                var containingNamespace = ImportCompletionItem.GetContainingNamespace(item);
-                if (!namespacesInScope.Contains(containingNamespace))
+                if (ShouldAddItem(item, namespacesInScope, aliasTargetNamespaceToTypeNameMap))
                 {
                     // We can return cached item directly, item's span will be fixed by completion service.
                     // On the other hand, because of this (i.e. mutating the  span of cached item for each run),
-                    // the provider can not be used as a service by components that might be run in parallel 
+                    // the provider can not be used as a service by components that might be run in parallel
                     // with completion, which would be a race.
                     completionContext.AddItem(item);
                     counter.ItemsCount++;
                 }
             }
+
+            static bool ShouldAddItem(
+                CompletionItem item,
+                HashSet<string> namespacesInScope,
+                MultiDictionary<string, string> aliasTargetNamespaceToTypeNameMap)
+            {
+                var containingNamespace = ImportCompletionItem.GetContainingNamespace(item);
+                // 1. if the namespace of the item is in scoop. Don't add the item
+                if (namespacesInScope.Contains(containingNamespace))
+                {
+                    return false;
+                }
+
+                // 2. If the item might be an alias target. First check if its namespace is in the map.
+                if (aliasTargetNamespaceToTypeNameMap.ContainsKey(containingNamespace))
```

Comment:
I'd suggest to check if the dictionary is empty just once before iterating though the items (i.e. no using alias in the document, which is usually the case). Basically have a check for a boolean value first to avoid any cost of calling `ContainsKey` (albeit small). 

Also you can just call `aliasTargetNamespaceToTypeNameMap[containingNamespace]` here, since the cost would be similar to containskey and our implementation would return an empty set if key can't be found. 

http://sourceroslyn.io/#Microsoft.CodeAnalysis/InternalUtilities/MultiDictionary.cs,225